### PR TITLE
Compute type based on value

### PR DESCRIPTION
**Pull request by arv**

### CODE REVIEW

**Comment by arv** [comment]
WIP

This passes go/types. I haven't tried to clean it up or tried it on other parts yet.

File: go/types/list_test.go [hunk 1]
```diff
@@ -151,16 +151,12 @@ func (suite *listTestSuite) TestMap() {
 	}
 }
 
-func TestListSuite1K(t *testing.T) {
```
Comment (arv) [inline go/types/list_test.go]: Why was this removed?

File: go/types/ref.go [hunk 2]
```diff
@@ -78,7 +79,7 @@ func (r Ref) WalkRefs(cb RefCallback) {
 }
 
 func (r Ref) typeOf() *Type {
-	return r.t
+	return MakeRefType(r.targetType)
```
Comment (arv) [inline go/types/ref.go]: This should be `makeCompoundType(RefKind, r.targetType)`

Comment (arv) [inline go/types/ref.go]: Done.

File: go/types/rolling_value_hasher.go [hunk 3]
```diff
@@ -35,8 +35,8 @@ func chunkingConfig() (pattern, window uint32) {
 func smallTestChunks() {
 	chunkConfigMu.Lock()
 	defer chunkConfigMu.Unlock()
-	chunkPattern = uint32(1<<7 - 1) // Avg Chunk Size of 128 bytes
-	chunkWindow = uint32(32)
+	chunkPattern = uint32(1<<8 - 1) // Avg Chunk Size of 256 bytes
```
Comment (arv) [inline go/types/rolling_value_hasher.go]: motivation?

File: go/types/struct.go [hunk 4]
```diff
@@ -16,56 +16,58 @@ import (
...
-func (fs structValueFields) Len() int           { return len(fs) }
-func (fs structValueFields) Swap(i, j int)      { fs[i], fs[j] = fs[j], fs[i] }
-func (fs structValueFields) Less(i, j int) bool { return fs[i].name < fs[j].name }
+	d.PanicIfFalse(len(s.fieldNames) == len(s.values))
 
-type Struct struct {
-	name   string
-	fields structValueFields
-	h      *hash.Hash
+	if len(s.fieldNames) == 0 {
```
Comment (arv) [inline go/types/struct.go]: Not sure this is worth it?

File: go/types/value.go [hunk 5]
```diff
@@ -35,11 +35,6 @@ type Value interface {
 	// chunked then this will return the refs of th sub trees of the prolly-tree.
 	WalkRefs(RefCallback)
 
-	// Type returns the type of the Noms value. All Noms values carry their runtime Noms type.
-	// DEPRECATED: https://github.com/attic-labs/noms/issues/3184
-	// Type() *Type
-
-	// Kind is the NomsKind describing the kind of value this is.
```
Comment (arv) [inline go/types/value.go]: Can we keep the godoc here?

Done.

File: go/types/value_decoder.go [hunk 6]
```diff
@@ -49,23 +50,26 @@ func (r *valueDecoder) readType() *Type {
...
+		return makeCompoundType(SetKind, r.readType(seenStructs))
 	case StructKind:
-		return r.readStructType()
+		return r.readStructType(seenStructs)
 	case UnionKind:
-		return r.readUnionType()
+		return r.readUnionType(seenStructs)
 	case CycleKind:
-		return MakeCycleType(r.readUint32())
+		name := r.readString()
```
Comment (arv) [inline go/types/value_decoder.go]: This changes `Cycle<0>` to `Cycle<Name>`. I'll have to update the code in a few more places to ensure consistency.

File: go/types/value_decoder.go [hunk 7]
```diff
@@ -155,47 +155,52 @@ func (r *valueDecoder) readValue() Value {
...
 		}
 
-		return newSet(r.readSetLeafSequence(t))
+		return newSet(r.readSetLeafSequence())
 	case StructKind:
-		return r.readStruct(t)
+		return r.readStruct()
 	case TypeKind:
-		return r.readType()
+		return r.readType(map[string]*Type{})
```
Comment (arv) [inline go/types/value_decoder.go]: Maybe we should use nil here instead?

File: go/types/value_decoder.go [hunk 8]
```diff
@@ -155,47 +155,52 @@ func (r *valueDecoder) readValue() Value {
...
 	case CycleKind, UnionKind, ValueKind:
-		d.Chk.Fail(fmt.Sprintf("A value instance can never have type %s", KindToString[t.TargetKind()]))
+		d.Chk.Fail(fmt.Sprintf("A value instance can never have type %s", KindToString[k]))
```
Comment (arv) [inline go/types/value_decoder.go]: I'm going to add `String()` to NomsKind. Too tired of guessing...

File: go/types/value_decoder.go [hunk 9]
```diff
@@ -205,27 +210,31 @@ func boolToUint32(b bool) uint32 {
...
+		t.Desc.(StructDesc).fields[i] = StructField{
 			r.readString(),
-			r.readType(),
+			r.readType(seenStructs),
 			r.readBool(),
 		}
 	}
 
-	return makeStructTypeQuickly(name, fields, checkKindNoValidate)
+	// Validate?
```
Comment (arv) [inline go/types/value_decoder.go]: We do not validate in read...

File: go/types/value_encoder.go [hunk 10]
```diff
@@ -183,24 +191,24 @@ func (w *valueEncoder) writeCycle(i uint32) {
...
+	seenStructs[name] = t
 
 	w.writeKind(StructKind)
-	desc := t.Desc.(StructDesc)
 	w.writeString(desc.Name)
-
 	w.writeUint32(uint32(desc.Len()))
 
 	for _, field := range desc.fields {
 		w.writeString(field.Name)
```
Comment (arv) [inline go/types/value_encoder.go]: Should we write the names together here too?

File: go/types/value_decoder.go (hunk 7 above)
Comment (arv) [inline go/types/value_decoder.go]: Answering myself again... Cannot be nil because we want to mutate it

File: go/types/list_test.go (hunk 1 above)
Comment (rafael-atticlabs) [inline go/types/list_test.go]: The 1K suites weren't chunking anymore so they were kind of pointless.

File: go/types/rolling_value_hasher.go (hunk 3 above)
Comment (rafael-atticlabs) [inline go/types/rolling_value_hasher.go]: i had to increase these to avoid the infinite chunking bug when in "smallTestChunks" mode.

one side effect of this is that we are once again chunking over type bytes in refs.

**Comment by arv** [comment]
@rafael-atticlabs This is now ready for review.

I have one more todo that I want to see if I can figure out (grouping cycle types in simplify).

Sorry about the size :'(

File: go/types/value_decoder.go (hunk 9 above)
Comment (rafael-atticlabs) [inline go/types/value_decoder.go]: remove comment.

File: go/types/value_encoder.go (hunk 10 above)
Comment (rafael-atticlabs) [inline go/types/value_encoder.go]: yeah, let's go ahead and do that now just to keep consistency. i think the same argument about compression efficiency *might* apply in some cases.

File: samples/go/csv/read.go [hunk 11]
```diff
@@ -113,7 +113,13 @@ func ReadToList(r *csv.Reader, structName string, headers []string, kinds KindSl
 		}
 
 		fields := readFieldsFromRow(row, headers, fieldOrder, kindMap)
-		valueChan <- types.NewStructWithType(t, fields)
+		data := make(types.StructData, len(fields))
```
Comment (rafael-atticlabs) [inline samples/go/csv/read.go]: this is going to regress csv reader perf. can you open a bug to new `NewStructWithFields`?

**Comment by arv** [comment]
Landing!!!
